src/backend/distributed/worker/worker_shard_visibility.c

```diff
@@ -460,8 +460,19 @@ FilterShardsFromPgclass(Node *node, void *context)
 			MemoryContext originalContext = MemoryContextSwitchTo(queryContext);
 
 			/* add NOT relation_is_a_known_shard(oid) to the security quals of the RTE */
-			rangeTableEntry->securityQuals =
-				list_make1(CreateRelationIsAKnownShardFilter(varno));
+			Node *oldQuals = query->jointree->quals;
+			if (oldQuals)
+			{
+				query->jointree->quals = (Node *) makeBoolExpr(
+					AND_EXPR,
+					list_make2(oldQuals, CreateRelationIsAKnownShardFilter(varno)),
+					-1);
+			}
+			else
+			{
+				query->jointree->quals = (Node *) CreateRelationIsAKnownShardFilter(
+					varno);
+			}
```

Comment:
This comes down to personal preference, but I'd like it better if we create the qual outside of a branch, and branch only to either _set_ the qual, or wrap it inside an and qual.

At the same time, we would possibly want a variation of `makeAndExpr` from gram.c (generated code so can't link - also static so we can't directly reuse) that can do the wrapping in an `And` condition, as well as folding it in a possibly already existing qual, although I think that optimization might be too over the top for _this_ usecase.

Given this is mostly about personal preference, feel free to ignore all or parts of this. However, given the empty quals branch is uncovered by tests, it would probably be moderately better to have it as simple as possible, eg, no function call, simply an assignment.

(bonus points for a quick add of a test that covers that case I guess :/ )

Reply:
didn't do the `makeAndExpr` as it indeed seems overkill here. But I did move the call outside the if and added a test.